pkg/build/api/validation/validation.go

```diff
@@ -145,23 +147,63 @@ func validateSource(input *buildapi.BuildSource) fielderrors.ValidationErrorList
 			allErrs = append(allErrs, validateGitSource(input.Git).Prefix("git")...)
 		}
 		if input.Dockerfile != nil {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("dockerfile", "", "may not be set when type is Git"))
+			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
+		}
+		if input.Binary != nil {
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("binary", "", "may not be set when type is Git"))
+		}
+	case buildapi.BuildSourceBinary:
+		if input.Binary == nil {
+			allErrs = append(allErrs, fielderrors.NewFieldRequired("binary"))
+		} else {
+			allErrs = append(allErrs, validateBinarySource(input.Binary).Prefix("binary")...)
+		}
+		if input.Dockerfile != nil {
+			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
+		}
+		if input.Git != nil {
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("git", "", "may not be set when type is Binary"))
```

Comment:
In my head they are conflicting - if you really need that case you can tar
your git repo and binary push.  That said, since at any time you can do a
binary build, you want to have a clear way to understand the impact.  I
feel like if you provide a directory, it conflicts with something else
providing a dir.  If you provide a file it doesn't conflict with a dir, and
vice versa.  Layering requires order decisions, and that scares me.

On Sep 28, 2015, at 4:38 PM, Ben Parees <notifications@github.com> wrote:

In pkg/build/api/validation/validation.go
<https://github.com/openshift/origin/pull/4819#discussion_r40559076>:

> +			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
> +		}
> +		if input.Binary != nil {
> +			allErrs = append(allErrs, fielderrors.NewFieldInvalid("binary", "", "may not be set when type is Git"))
> +		}
> +	case buildapi.BuildSourceBinary:
> +		if input.Binary == nil {
> +			allErrs = append(allErrs, fielderrors.NewFieldRequired("binary"))
> +		} else {
> +			allErrs = append(allErrs, validateBinarySource(input.Binary).Prefix("binary")...)
> +		}
> +		if input.Dockerfile != nil {
> +			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
> +		}
> +		if input.Git != nil {
> +			allErrs = append(allErrs, fielderrors.NewFieldInvalid("git", "", "may not be set when type is Binary"))

why not? can't i supply some binary artifacts and some git source, together?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4819/files#r40559076>.